File: 00015_3sum/15.三数之和.py
#
# @lc app=leetcode.cn id=15 lang=python3
#
# [15] 三数之和
#

# @lc code=start


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        result = []
        if len(nums) < 3:
            return result
        
        nums.sort()
        for i in range(len(nums)):
            if nums[i] > 0: return result
            if i>0 and nums[i] == nums[i-1]: continue

            left = i + 1
            right = len(nums) - 1
            while left < right:
                sum_result = nums[i] + nums[left] + nums[right]
                if sum_result == 0:
                    result.append([nums[i] ,nums[left] ,nums[right]])
                    while left<right and nums[left] == nums[left+1]: left+=1
                    while left<right and nums[right] == nums[right-1]: right-=1

                    left+=1
                    right-=1
                elif sum_result < 0:
                    left+=1
                elif sum_result > 0:
                    right-=1
        return result
# @lc code=end

File: 00015_3sum/15.三数之和.java
/*
 * @lc app=leetcode.cn id=15 lang=java
 *
 * [15] 三数之和
 */

// @lc code=start
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums);

        for (int i=0; i<nums.length;i++) {
            if (nums[i] > 0) {
                break;
            }

            if (i>0 && nums[i] == nums[i-1]) {
                continue;
            }

            int left = i + 1;
            int right = nums.length - 1;
            while (left < right) {
                int sumRes = nums[i] + nums[left] + nums[right];
                if (sumRes == 0) {
                    result.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    while (left < right && nums[left] == nums[left+1]) left++;
                    while (left < right && nums[right] == nums[right-1]) right--;

                    right--;
                    left++;
                } else if (sumRes > 0) {
                    right--;
                } else if (sumRes < 0) {
                    left++;
                }
            }
        }

        return result;
    }
}
// @lc code=end

